### PR TITLE
Vs backend compiler selection

### PR DESCRIPTION
**Pull request by dcbaker**
This little series fixes ICL (which was broken by a refactor), and adds support for using clang-cl and icl for the visual studio backend.

@jon-turney, don't know if you're interested or not.

### CODE REVIEW

**Comment by jpakkane** [comment]
```
  File "D:\a\1\s\mesonbuild\interpreter.py", line 2678, in func_project
    self.set_backend()
  File "D:\a\1\s\mesonbuild\interpreter.py", line 2595, in set_backend
    self.backend = backends.get_backend_from_name(backend, self.build)
  File "D:\a\1\s\mesonbuild\backend\backends.py", line 144, in get_backend_from_name
    return vs2015backend.Vs2015Backend(build)
  File "D:\a\1\s\mesonbuild\backend\vs2015backend.py", line 25, in __init__
    c = list(comps.values())[0]
IndexError: list index out of range
```

File: docs/markdown/Using-with-Visual-Studio.md [hunk 1]
```diff
@@ -14,3 +14,15 @@ In order to generate Visual Studio projects, Meson needs to know the settings of
...
 
 This assumes Python3 is in your `PATH`, which is highly recommended.
+
+# Using Clang-CL with Visual Studio
+
+You will first need to get a copy of llvm+clang for Windows, such versions are available from a number of sources, including the llvm website. Then you will need the [llvm toolset extension for visual studio](https://marketplace.visualstudio.com/items?itemName=LLVMExtensions.llvm-toolchain). You then need to either use a native file or `set CC=clang-cl`, and `set CXX=clang-cl` to use those compilers, meson will do the rest.
+
+This only works with visual studio 2017 and 2019.
+
+There is currently not support in meson for clang/c2
```
Comment (TheQwertiest) [inline docs/markdown/Using-with-Visual-Studio.md]: s/not/no support

File: mesonbuild/backend/vs2017backend.py [hunk 2]
```diff
@@ -22,8 +22,17 @@ class Vs2017Backend(Vs2010Backend):
...
-        self.platform_toolset = 'v141'
         self.vs_version = '2017'
+        comps = self.environment.coredata.compilers
+        if all(c.id == 'clang-cl' for c in comps.values()):
+            self.platform_toolset = 'llvm'
+        elif all(c.id == 'intel-cl' for c in comps.values()):
+            c = list(comps.values())[0]
+            if c.version.startswith('19'):
+                self.platform_toolset = 'Intel C++ Compiler 19.0'
+            # We don't have support for versions nolder than 2019 right now.
```
Comment (TheQwertiest) [inline mesonbuild/backend/vs2017backend.py]: s/nolder/older

File: docs/markdown/Using-with-Visual-Studio.md [hunk 3]
```diff
@@ -14,3 +14,15 @@ In order to generate Visual Studio projects, Meson needs to know the settings of
 If you wish to use the Ninja backend instead of vs2015, pass `--backend ninja`. At the time of writing the Ninja backend is more mature than the VS backend so you might want to use it for serious work.
 
 This assumes Python3 is in your `PATH`, which is highly recommended.
+
+# Using Clang-CL with Visual Studio
+
+You will first need to get a copy of llvm+clang for Windows, such versions are available from a number of sources, including the llvm website. Then you will need the [llvm toolset extension for visual studio](https://marketplace.visualstudio.com/items?itemName=LLVMExtensions.llvm-toolchain). You then need to either use a native file or `set CC=clang-cl`, and `set CXX=clang-cl` to use those compilers, meson will do the rest.
+
```
Comment (jon-turney) [inline docs/markdown/Using-with-Visual-Studio.md]: Probably should say something more about 'Use a native file to specify the compiler'?
Also wrap lines?

File: docs/markdown/snippets/vstoolsets.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+## Support for non-msvc compilers with visual studio backends
+
+Currently using any VisualStudio Backend will result in MSVC being used to
```
Comment (jon-turney) [inline docs/markdown/snippets/vstoolsets.md]: 'With previous versions' is possibly clearer than 'Currently'

File: docs/markdown/snippets/vstoolsets.md [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+compile, even if a different compiler has been set via environment variables.
+This has been fixed so that Clang-CL can also be used. Support for the Intel ICL compiler is also available.
```
Comment (jon-turney) [inline docs/markdown/snippets/vstoolsets.md]: Why not say "Clang-Cl or Intel ICL can also be used"?

File: mesonbuild/backend/vs2015backend.py [hunk 6]
```diff
@@ -19,5 +19,12 @@ class Vs2015Backend(Vs2010Backend):
...
         self.name = 'vs2015'
-        self.platform_toolset = 'v140'
         self.vs_version = '2015'
+        comps = self.environment.coredata.compilers
+        if all(c.id == 'intel-cl' for c in comps.values()):
+            c = list(comps.values())[0]
+            if c.version.startswith('19'):
+                self.platform_toolset = 'Intel C++ Compiler 19.0'
+            # We don't have support for versions nolder than 2019 right now.
+        if self.platform_toolset is None:
```
Comment (jon-turney) [inline mesonbuild/backend/vs2015backend.py]: Unclear to me how this is going to handle combinations like 'CC=icl, CXX=cl'.  Are those unsupported?  Should that be mentioned in snippet?

**Comment by jon-turney** [comment]
> @jon-turney, don't know if you're interested or not.

My interest in VS is mainly limited to making sure fixes to it don't break stuff I actually care about :)

Maybe @nioncode is more interested in it than I am

**Comment by nioncode** [comment]
> Maybe @nioncode is more interested in it than I am

Unfortunately we do not use meson at work at the moment, so I'm not actively using it anymore.

File: mesonbuild/backend/vs2015backend.py (hunk 6 above)
Comment (dcbaker) [reply mesonbuild/backend/vs2015backend.py]: I've added an exception for that case, as well as one for older versions of ICL.

Comment (dcbaker) [reply mesonbuild/backend/vs2015backend.py]: I've also mentioned that in the docs.

Comment (dcbaker) [reply mesonbuild/backend/vs2015backend.py]: I've also mentioned that in the docs.

File: mesonbuild/backend/vs2017backend.py [hunk 7]
```diff
@@ -22,8 +22,20 @@ class Vs2017Backend(Vs2010Backend):
...
+        if comps:
+            if all(c.id == 'clang-cl' for c in comps.values()):
+                self.platform_toolset = 'llvm'
+            elif all(c.id == 'intel-cl' for c in comps.values()):
+                c = list(comps.values())[0]
+                if c.version.startswith('19'):
+                    self.platform_toolset = 'Intel C++ Compiler 19.0'
+                else:
+                    # We don't have support for versions older than 2019 right now.
+                    raise MesonException('There is currently no support for ICL before 19, patches welcome.')
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2017backend.py]: __[Flake8]__ [F821] undefined name 'MesonException' ([link](https://sider.review/gh/repos/19784232/pulls/5448#issue-4186287))

<sub>posted by Sider</sub>

File: mesonbuild/backend/vs2015backend.py [hunk 8]
```diff
@@ -19,5 +19,14 @@ class Vs2015Backend(Vs2010Backend):
...
-        self.platform_toolset = 'v140'
         self.vs_version = '2015'
+        comps = self.environment.coredata.compilers
+        if comps and all(c.id == 'intel-cl' for c in comps.values()):
+            c = list(comps.values())[0]
+            if c.version.startswith('19'):
+                self.platform_toolset = 'Intel C++ Compiler 19.0'
+            else:
+                # We don't have support for versions older than 2019 right now.
+                raise MesonException('There is currently no support for ICL before 19, patches welcome.')
```
Comment (ignatenkobrain) [inline mesonbuild/backend/vs2015backend.py]: __[Flake8]__ [F821] undefined name 'MesonException' ([link](https://sider.review/gh/repos/19784232/pulls/5448#issue-4186286))

<sub>posted by Sider</sub>

**Comment by dcbaker** [comment]
Okay, I think I've addressed all of the comments now.

File: docs/markdown/Using-with-Visual-Studio.md (hunk 3 above)
Comment (dcbaker) [reply docs/markdown/Using-with-Visual-Studio.md]: Added a link to the native files documentation and wrapped the whole thing.

**Comment by jpakkane** [comment]
Tests still fail. For a different reason, though.

```
RuntimeError: Failed to configure 'test cases/common/1 trivial':
Traceback (most recent call last):
  File "D:\a\1\s\mesonbuild\mesonmain.py", line 127, in run
    return options.run_func(options)
  File "D:\a\1\s\mesonbuild\msetup.py", line 240, in run
    app.generate()
  File "D:\a\1\s\mesonbuild\msetup.py", line 158, in generate
    self._generate(env)
  File "D:\a\1\s\mesonbuild\msetup.py", line 174, in _generate
    intr = interpreter.Interpreter(b)
  File "D:\a\1\s\mesonbuild\interpreter.py", line 2072, in __init__
    self.parse_project()
  File "D:\a\1\s\mesonbuild\interpreterbase.py", line 392, in parse_project
    self.evaluate_codeblock(self.ast, end=1)
  File "D:\a\1\s\mesonbuild\interpreterbase.py", line 431, in evaluate_codeblock
    raise e
  File "D:\a\1\s\mesonbuild\interpreterbase.py", line 425, in evaluate_codeblock
    self.evaluate_statement(cur)
  File "D:\a\1\s\mesonbuild\interpreterbase.py", line 436, in evaluate_statement
    return self.function_call(cur)
  File "D:\a\1\s\mesonbuild\interpreterbase.py", line 773, in function_call
    return func(node, posargs, kwargs)
  File "D:\a\1\s\mesonbuild\interpreterbase.py", line 143, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "D:\a\1\s\mesonbuild\interpreterbase.py", line 174, in wrapped
    return f(*wrapped_args, **wrapped_kwargs)
  File "D:\a\1\s\mesonbuild\interpreter.py", line 2720, in func_project
    self.set_backend()
  File "D:\a\1\s\mesonbuild\interpreter.py", line 2637, in set_backend
    self.backend = backends.get_backend_from_name(backend, self.build)
  File "D:\a\1\s\mesonbuild\backend\backends.py", line 150, in get_backend_from_name
    return vs2019backend.Vs2019Backend(build)
  File "D:\a\1\s\mesonbuild\backend\vs2019backend.py", line 23, in __init__
    super().__init__(build)
  File "D:\a\1\s\mesonbuild\backend\vs2010backend.py", line 100, in __init__
    ids = [comps[0].id for c in comps.values()]
AttributeError: 'PerMachine' object has no attribute 'values'
```

**Comment by dcbaker** [comment]
Ah rebases :(

**Comment by dcbaker** [comment]
@jpakkane, do we support a build machine compiler with the vs backend, or is that just ignored?

**Comment by jpakkane** [comment]
I don't think anyone has tried that with MSVC. VS is unfortunate because all its compilers are called `cl.exe` so you can't really invoke many of them from one command prompt.

**Comment by dcbaker** [comment]
I assume you'd have to do something tricky like wrap them in a python script, but if that's the case I'm just going to look at the `host` compilers and not bother with the `build` compilers. It probably can be worked around, but if it didn't work before...

**Comment by jpakkane** [comment]
192 fails with:

```
Project "D:\a\1\s\b b67ca84f90\args flattening.sln" (1) is building "D:\a\1\s\b b67ca84f90\REGEN.vcxproj" (2) on node 1 (default targets).
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\MSBuild\Microsoft\VC\v160\Microsoft.CppBuild.targets(379,5): error MSB8020: The build tools for llvm (Platform Toolset = 'llvm') cannot be found. To build using the llvm build tools, please install llvm build tools.  Alternatively, you may upgrade to the current Visual Studio tools by selecting the Project menu or right-click the solution, and then selecting "Retarget solution". [D:\a\1\s\b b67ca84f90\REGEN.vcxproj]
Done Building Project "D:\a\1\s\b b67ca84f90\REGEN.vcxproj" (default targets) -- FAILED.
Done Building Project "D:\a\1\s\b b67ca84f90\args flattening.sln" (default targets) -- FAILED.
```

**Comment by dcbaker** [comment]
I just noticed that I can get vs2019 through work now, I'm installing it now and I'll get that sorted.

**Comment by jpakkane** [comment]
The error message is still the same. The setup is for `vs`, not `llvm`. It should not be trying to use llvm tools in this case, right?

**Comment by dcbaker** [comment]
Right, it shouldn't try to use llvm unless one of the compilers is clang-cl. I need to look at my machine I left running after work last night and see if I was able to reproduce.

**Comment by dcbaker** [comment]
python's `all()` has rather surprising behavior. It turns out that `all([]) == True`, which is not at all what I expected to happen, so if you have a project with no compilers used, llvm will be selected for 2017 and 2019, while intel will be selected for 2015. If this comes back green I'll clean up the patches.

**Comment by dcbaker** [comment]
Okay, patches all cleaned up, hopefully I didn't screw up the rebase.

**Comment by dcbaker** [comment]
@jpakkane, all failures are now the expected pch msys ones.

**Pull request merged**